CalibMuon/DTCalibration/plugins/DTResidualHistory.cc

```diff
@@ -0,0 +1,308 @@
+	      for(; layer_it != layer_end; ++layer_it) { 
+		 DTLayerId layerId = (*layer_it)->id();
+		 bookHistos(layerId,run);
+              }
+           }
+        }
+      }    
+
+    std::stringstream runStr; runStr << run;
+    TDirectory* baseDir = rootFile_->GetDirectory(("Run"+runStr.str()).c_str());
+    if(!baseDir) baseDir = rootFile_->mkdir(("Run"+runStr.str()).c_str());
+    baseDir->cd();
+    histoResLs = new TH2F("histoResLs","Residuals vs Lumisection",100,0,10000,100,-1,1);
+    std::vector<TH2F*> histosTH2F;
+    histosTH2F.push_back(histoResLs);
+    histoMapTH2F_[slIdToFill] = histosTH2F;
+    rootFile_->cd();
+  
+    lastrun=run;
+  }
+
+  // Get the 4D rechits from the event
+  edm::Handle<DTRecSegment4DCollection> segment4Ds;
+  event.getByLabel(segment4DLabel_, segment4Ds);
+ 
+  // Loop over segments by chamber
+  DTRecSegment4DCollection::id_iterator chamberIdIt;
+  for(chamberIdIt = segment4Ds->id_begin(); chamberIdIt != segment4Ds->id_end(); ++chamberIdIt){
+
+     // Get the range for the corresponding ChamberId
+     DTRecSegment4DCollection::range range = segment4Ds->get((*chamberIdIt));
+     // Loop over the rechits of this DetUnit
+     for(DTRecSegment4DCollection::const_iterator segment  = range.first;
+                                                  segment != range.second; ++segment){
+
+        if( !select_(*segment, event, setup) ) { segmbad++; continue; }
+        segmok++;
+
+        // Get all 1D RecHits at step 3 within the 4D segment
+        std::vector<DTRecHit1D> recHits1D_S3;
+  
+        if( (*segment).hasPhi() ){
+           const DTChamberRecSegment2D* phiSeg = (*segment).phiSegment();
+           const std::vector<DTRecHit1D>& phiRecHits = phiSeg->specificRecHits();
+           std::copy(phiRecHits.begin(), phiRecHits.end(), back_inserter(recHits1D_S3));
+        }
+
+        if( (*segment).hasZed() ){
+           const DTSLRecSegment2D* zSeg = (*segment).zSegment();
+           const std::vector<DTRecHit1D>& zRecHits = zSeg->specificRecHits();
+           std::copy(zRecHits.begin(), zRecHits.end(), back_inserter(recHits1D_S3));
+        }
+
+        // Loop over 1D RecHit inside 4D segment
+        for(std::vector<DTRecHit1D>::const_iterator recHit1D = recHits1D_S3.begin();
+                                                    recHit1D != recHits1D_S3.end(); ++recHit1D) {
+           const DTWireId wireId = recHit1D->wireId();
+           float residualOnDistance = DTRecHitSegmentResidual().compute(dtGeom_,*recHit1D,*segment);
+           fillHistos(wireId.superlayerId(), residualOnDistance);
+           if (wireId.superlayerId() == slIdToFill) histoResLs->Fill(event.id().luminosityBlock(),residualOnDistance);
+           if(detailedAnalysis_) fillHistos(wireId.layerId(), residualOnDistance);
+        }
+     }
+  }
+}
+
+float DTResidualHistory::segmentToWireDistance(const DTRecHit1D& recHit1D, const DTRecSegment4D& segment){
```

Comment:
it looks like this function already exists in CMSSW - please make a helper class to share code.
https://github.com/cms-sw/cmssw/blob/09c3fce6626f70fd04223e7dacebf0b485f73f54/CalibMuon/DTCalibration/plugins/DTResidualCalibration.h#L43